firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -4578,22 +4578,324 @@ public void testGetWaitsForConnection() throws DatabaseException, InterruptedExc
+    String writeKey = writeNode.getKey();
+    DatabaseReference readNode = readDb.getReference().child(writeKey);
+
+    try {
+      await(writeNode.setValue(val));
+      DatabaseReference topLevelNode = readDb.getReference();
+      ReadFuture readFuture =
+          new ReadFuture(
+              topLevelNode,
+              events -> {
+                assertEquals(1, events.size());
+                DataSnapshot childNode = events.get(0).getSnapshot().child(writeKey);
+                assertEquals(val, childNode.getValue());
+                return true;
+              });
+      readFuture.timedGet();
+      DataSnapshot snapshot = null;
+      snapshot = await(readNode.get());
+      assertEquals(val, Objects.requireNonNull(snapshot.getValue()));
+    } catch (TestFailure | TimeoutException | ExecutionException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Test
+  public void testGetResolvesToCacheWhenOnlineAndSameLevelListener()
+      throws DatabaseException, InterruptedException, ExecutionException, TestFailure,
+          TimeoutException {
+    FirebaseApp writeApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    // To ensure that we don't read the cached results, we need a separate app.
+    FirebaseApp readApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+    FirebaseDatabase readDb = FirebaseDatabase.getInstance(readApp);
+    long val = 34;
+    String writeKey = Objects.requireNonNull(writeDb.getReference().push().getKey());
+    DatabaseReference writeNode = writeDb.getReference().child(writeKey);
+    await(writeNode.setValue(val));
+    DatabaseReference readNode = readDb.getReference().child(writeKey);
+    new ReadFuture(
+            readNode,
+            events -> {
+              assertEquals(1, events.size());
+              assertEquals(val, events.get(0).getSnapshot().getValue());
+              return true;
+            })
+        .timedGet();
+
+    try {
+      DataSnapshot snapshot = await(readNode.get());
+      assertEquals(val, snapshot.getValue());
+      // need to rewrite it
+    } catch (ExecutionException e) {
+      fail("get threw an exception: " + e);
+    }
+  }
+
+  @Test
+  public void testGetResolvesToServerCacheWhenListenerIsAvailable()
+      throws DatabaseException, InterruptedException {
+    // To ensure that we don't read the cached results, we need a separate app.
+    FirebaseApp app =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase db = FirebaseDatabase.getInstance(app);
+    db.setPersistenceEnabled(false);
+    long val = 34;
+    DatabaseReference node = db.getReference().push();
+
+    try {
+      await(node.setValue(val));
+      Semaphore semaphore = new Semaphore(0);
+      ValueEventListener listener =
```

Comment:
Is there a reason we can't use `ReadFuture` here?

Reply:
ReadFuture will remove the listener after the assertion, and if that happens, it takes the cache along with it. This is so that we can make the assertion, and then remove the listener at a later time.